File: projet-universitaire/backend/src/main/java/com/supcom/projetuniversitaire/service/ProjetService.java
package com.supcom.projetuniversitaire.service;

import com.supcom.projetuniversitaire.exception.ResourceNotFoundException;
import com.supcom.projetuniversitaire.model.Projet;
import com.supcom.projetuniversitaire.model.Tache;
import com.supcom.projetuniversitaire.repository.ProjetRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * ============================================================
 * SERVICE : ProjetService
 * ============================================================
 * Couche de logique métier pour la gestion des projets universitaires.
 * Cette classe fait le lien entre le Controller (HTTP) et le Repository (DB).
 *
 * Responsabilités :
 *   - Valider et traiter les données avant de les sauvegarder
 *   - Calculer automatiquement l'avancement (%)
 *   - Vérifier les indicateurs de retard
 *   - Gérer les tâches (ajout, modification, suppression)
 * ============================================================
 */
@Service
public class ProjetService {

    // Injection du repository pour accéder à MongoDB
    @Autowired
    private ProjetRepository projetRepository;

    // ============================================================
    // OPÉRATIONS CRUD SUR LES PROJETS
    // ============================================================

    /**
     * Récupère tous les projets de la base de données.
     * Met à jour les indicateurs de retard avant de retourner.
     *
     * @return liste de tous les projets
     */
    public List<Projet> getTousLesProjets() {
        List<Projet> projets = projetRepository.findAll();
        // Mettre à jour le statut de retard pour chaque projet
        projets.forEach(p -> {
            p.verifierRetard();
            // Mettre à jour les tâches en retard
            if (p.getTaches() != null) {
                p.getTaches().forEach(Tache::verifierRetard);
            }
        });
        return projets;
    }

    /**
     * Récupère un projet par son identifiant MongoDB.
     *
     * @param id l'identifiant du projet
     * @return le projet trouvé
     * @throws ResourceNotFoundException si le projet n'existe pas
     */
    public Projet getProjetParId(String id) {
        Projet projet = projetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "Projet non trouvé avec l'identifiant : " + id));

        // Mettre à jour l'avancement et le statut avant de retourner
        projet.calculerAvancement();
        projet.verifierRetard();

        return projet;
    }

    /**
     * Crée un nouveau projet et le sauvegarde en base.
     * Initialise l'avancement à 0% et le statut à "EN_COURS".
     *
     * @param projet le projet à créer
     * @return le projet créé avec son ID généré par MongoDB
     */
    public Projet creerProjet(Projet projet) {
        // Initialisation de l'avancement
        projet.calculerAvancement();
        // Vérification du statut initial
        projet.verifierRetard();
        // Sauvegarde en base MongoDB
        return projetRepository.save(projet);
    }

    /**
     * Met à jour un projet existant.
     * Recalcule automatiquement l'avancement après la mise à jour.
     *
     * @param id l'identifiant du projet à mettre à jour
     * @param projetDetails les nouvelles données du projet
     * @return le projet mis à jour
     * @throws ResourceNotFoundException si le projet n'existe pas
     */
    public Projet mettreAJourProjet(String id, Projet projetDetails) {
        // Vérifier que le projet existe
        Projet projetExistant = getProjetParId(id);

        // Mettre à jour les champs modifiables
        projetExistant.setTitre(projetDetails.getTitre());
        projetExistant.setDescription(projetDetails.getDescription());
        projetExistant.setMatiere(projetDetails.getMatiere());
        projetExistant.setDateLimite(projetDetails.getDateLimite());
        projetExistant.setMembres(projetDetails.getMembres());

        // Recalculer l'avancement après modification
        projetExistant.calculerAvancement();
        projetExistant.verifierRetard();

        // Sauvegarder les modifications
        return projetRepository.save(projetExistant);
    }

    /**
     * Supprime un projet par son identifiant.
     *
     * @param id l'identifiant du projet à supprimer
     * @throws ResourceNotFoundException si le projet n'existe pas
     */
    public void supprimerProjet(String id) {
        // Vérifier que le projet existe avant de supprimer
        Projet projet = getProjetParId(id);
        projetRepository.delete(projet);
    }

    // ============================================================
    // OPÉRATIONS SUR LES TÂCHES D'UN PROJET
    // ============================================================

    /**
     * Ajoute une nouvelle tâche à un projet existant.
     * Recalcule l'avancement après l'ajout.
     *
     * @param projetId l'identifiant du projet
     * @param tache    la tâche à ajouter
     * @return le projet mis à jour avec la nouvelle tâche
     */
    public Projet ajouterTache(String projetId, Tache tache) {
        Projet projet = getProjetParId(projetId);

        // Vérifier le retard de la nouvelle tâche
        tache.verifierRetard();

        // Ajouter la tâche à la liste
        projet.getTaches().add(tache);

        // Recalculer l'avancement global du projet
        projet.calculerAvancement();
        projet.verifierRetard();

        return projetRepository.save(projet);
    }

    /**
     * Met à jour le statut d'une tâche (TO_DO → DOING → DONE).
     * Recalcule automatiquement l'avancement du projet parent.
     *
     * @param projetId l'identifiant du projet
     * @param tacheId  l'identifiant de la tâche à modifier
     * @param nouveauStatut le nouveau statut ("TO_DO", "DOING", "DONE")
     * @return le projet mis à jour
     */
    public Projet mettreAJourStatutTache(String projetId, String tacheId, String nouveauStatut) {
        Projet projet = getProjetParId(projetId);

        // Trouver et mettre à jour la tâche concernée
        projet.getTaches().stream()
                .filter(t -> t.getId().equals(tacheId))
                .findFirst()
                .ifPresent(tache -> {
                    tache.setStatut(nouveauStatut);
                    tache.verifierRetard(); // Recalculer le retard après changement de statut
                });

        // Recalculer l'avancement global après le changement de statut
        projet.calculerAvancement();
        projet.verifierRetard();

        return projetRepository.save(projet);
    }

    /**
     * Supprime une tâche d'un projet.
     * Recalcule l'avancement après suppression.
     *
     * @param projetId l'identifiant du projet
     * @param tacheId  l'identifiant de la tâche à supprimer
     * @return le projet mis à jour
     */
    public Projet supprimerTache(String projetId, String tacheId) {
        Projet projet = getProjetParId(projetId);

        // Supprimer la tâche correspondante de la liste
        projet.getTaches().removeIf(t -> t.getId().equals(tacheId));

        // Recalculer l'avancement après suppression
        projet.calculerAvancement();
        projet.verifierRetard();

        return projetRepository.save(projet);
    }

    // ============================================================
    // OPÉRATIONS DE RECHERCHE ET FILTRAGE
    // ============================================================

    /**
     * Filtre les projets par statut.
     *
     * @param statut le statut à filtrer ("EN_COURS", "TERMINÉ", "EN_RETARD")
     * @return liste filtrée des projets
     */
    public List<Projet> getProjetsByStatut(String statut) {
        return projetRepository.findByStatut(statut);
    }

    /**
     * Ajoute un membre à un projet existant.
     *
     * @param projetId l'identifiant du projet
     * @param membre   le nom/email du membre à ajouter
     * @return le projet mis à jour
     */
    public Projet ajouterMembre(String projetId, String membre) {
        Projet projet = getProjetParId(projetId);

        // Vérifier que le membre n'est pas déjà dans le groupe
        if (!projet.getMembres().contains(membre)) {
            projet.getMembres().add(membre);
            return projetRepository.save(projet);
        }
        return projet;
    }
}

File: projet-universitaire/frontend/src/app/services/projet.service.ts
/**
 * ============================================================
 * SERVICE : ProjetService (Angular)
 * ============================================================
 * Service Angular pour communiquer avec l'API REST Spring Boot.
 * Utilise HttpClient pour envoyer des requêtes HTTP (GET, POST, PUT, DELETE).
 *
 * URL de base de l'API : http://localhost:8080/api/projets
 *
 * Ce service est injecté dans les composants Angular qui ont
 * besoin d'accéder aux données des projets.
 * ============================================================
 */
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Projet } from '../models/projet.model';
import { Tache } from '../models/tache.model';

@Injectable({
  providedIn: 'root' // Service disponible globalement dans toute l'application
})
export class ProjetService {

  /** URL de base de l'API backend Spring Boot */
  private apiUrl = 'http://localhost:8080/api/projets';

  /** Injection du client HTTP Angular */
  private http = inject(HttpClient);

  // ============================================================
  // MÉTHODES CRUD POUR LES PROJETS
  // ============================================================

  /**
   * Récupère la liste complète de tous les projets.
   * Correspond à : GET /api/projets
   *
   * @returns Observable<Projet[]> flux de données avec la liste des projets
   */
  getTousLesProjets(): Observable<Projet[]> {
    return this.http.get<Projet[]>(this.apiUrl);
  }

  /**
   * Récupère un projet spécifique par son identifiant.
   * Correspond à : GET /api/projets/{id}
   *
   * @param id l'identifiant MongoDB du projet
   * @returns Observable<Projet> flux contenant le projet
   */
  getProjetParId(id: string): Observable<Projet> {
    return this.http.get<Projet>(`${this.apiUrl}/${id}`);
  }

  /**
   * Crée un nouveau projet universitaire.
   * Correspond à : POST /api/projets
   *
   * @param projet les données du nouveau projet
   * @returns Observable<Projet> le projet créé avec son ID généré
   */
  creerProjet(projet: Projet): Observable<Projet> {
    return this.http.post<Projet>(this.apiUrl, projet);
  }

  /**
   * Met à jour un projet existant.
   * Correspond à : PUT /api/projets/{id}
   *
   * @param id      l'identifiant du projet à modifier
   * @param projet  les nouvelles données du projet
   * @returns Observable<Projet> le projet mis à jour
   */
  mettreAJourProjet(id: string, projet: Projet): Observable<Projet> {
    return this.http.put<Projet>(`${this.apiUrl}/${id}`, projet);
  }

  /**
   * Supprime un projet par son identifiant.
   * Correspond à : DELETE /api/projets/{id}
   *
   * @param id l'identifiant du projet à supprimer
   * @returns Observable<any> réponse de confirmation
   */
  supprimerProjet(id: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }

  // ============================================================
  // MÉTHODES POUR LA GESTION DES TÂCHES (KANBAN)
  // ============================================================

  /**
   * Ajoute une nouvelle tâche à un projet existant.
   * Correspond à : POST /api/projets/{id}/taches
   *
   * @param projetId l'identifiant du projet parent
   * @param tache    la tâche à ajouter
   * @returns Observable<Projet> le projet mis à jour avec la nouvelle tâche
   */
  ajouterTache(projetId: string, tache: Tache): Observable<Projet> {
    return this.http.post<Projet>(`${this.apiUrl}/${projetId}/taches`, tache);
  }

  /**
   * Met à jour le statut d'une tâche (déplacement Kanban).
   * Correspond à : PUT /api/projets/{id}/taches/{tacheId}/statut
   *
   * @param projetId      l'identifiant du projet
   * @param tacheId       l'identifiant de la tâche
   * @param nouveauStatut le nouveau statut : 'TO_DO', 'DOING', 'DONE'
   * @returns Observable<Projet> le projet mis à jour avec l'avancement recalculé
   */
  mettreAJourStatutTache(projetId: string, tacheId: string, nouveauStatut: string): Observable<Projet> {
    return this.http.put<Projet>(
      `${this.apiUrl}/${projetId}/taches/${tacheId}/statut`,
      { statut: nouveauStatut }
    );
  }

  /**
   * Supprime une tâche d'un projet.
   * Correspond à : DELETE /api/projets/{id}/taches/{tacheId}
   *
   * @param projetId l'identifiant du projet
   * @param tacheId  l'identifiant de la tâche à supprimer
   * @returns Observable<Projet> le projet mis à jour
   */
  supprimerTache(projetId: string, tacheId: string): Observable<Projet> {
    return this.http.delete<Projet>(`${this.apiUrl}/${projetId}/taches/${tacheId}`);
  }

  // ============================================================
  // MÉTHODES POUR LA GESTION DES MEMBRES DU GROUPE
  // ============================================================

  /**
   * Ajoute un membre au groupe d'un projet.
   * Correspond à : POST /api/projets/{id}/membres
   *
   * @param projetId l'identifiant du projet
   * @param membre   le nom ou email du membre à ajouter
   * @returns Observable<Projet> le projet mis à jour
   */
  ajouterMembre(projetId: string, membre: string): Observable<Projet> {
    return this.http.post<Projet>(`${this.apiUrl}/${projetId}/membres`, { membre });
  }

  /**
   * Filtre les projets par statut.
   * Correspond à : GET /api/projets/statut/{statut}
   *
   * @param statut 'EN_COURS', 'TERMINÉ', ou 'EN_RETARD'
   * @returns Observable<Projet[]> liste filtrée
   */
  getProjetsByStatut(statut: string): Observable<Projet[]> {
    return this.http.get<Projet[]>(`${this.apiUrl}/statut/${statut}`);
  }
}
